pkg/utils/addons/addons.go

```diff
@@ -0,0 +1,166 @@
+	if !found || err != nil {
+		return unstructured.Unstructured{}, err
+	}
+	taskDeployMap := taskDeploy.([]interface{})[2].(map[string]interface{})
+	taskDeployMap["taskRef"] = map[string]interface{}{"name": "openshift-client", "kind": "ClusterTask"}
+	taskDeployMap["runAfter"] = []interface{}{"build"}
+	taskDeployMap["params"] = []interface{}{
+		map[string]interface{}{"name": "ARGS", "value": []interface{}{"rollout", "status", "dc/$(params.APP_NAME)"}},
+	}
+
+	return newTempRes, nil
+}
+
+func (dp *deploymentPipeline) generate(pipeline unstructured.Unstructured) (unstructured.Unstructured, error) {
+	newTempRes := unstructured.Unstructured{}
+	pipeline.DeepCopyInto(&newTempRes)
+	labels := newTempRes.GetLabels()
+	labels[flag.LabelEnvironmentType] = dp.environment
+	newTempRes.SetLabels(labels)
+	name := newTempRes.GetName()
+	newTempRes.SetName(name + dp.nameSuffix)
+
+	taskDeploy, found, err := unstructured.NestedFieldNoCopy(newTempRes.Object, "spec", "tasks")
+	if !found || err != nil {
+		return unstructured.Unstructured{}, err
+	}
+	taskDeployMap := taskDeploy.([]interface{})[2].(map[string]interface{})
+	taskDeployMap["taskRef"] = map[string]interface{}{"name": "openshift-client", "kind": "ClusterTask"}
+	taskDeployMap["runAfter"] = []interface{}{"build"}
+	taskDeployMap["params"] = []interface{}{
+		map[string]interface{}{"name": "SCRIPT", "value": "kubectl $@"},
+		map[string]interface{}{"name": "ARGS", "value": []interface{}{"rollout", "status", "deploy/$(params.APP_NAME)"}},
+	}
+	return newTempRes, nil
+}
+
+func (kp *knativePipeline) generate(pipeline unstructured.Unstructured) (unstructured.Unstructured, error) {
+	newTempRes := unstructured.Unstructured{}
+	pipeline.DeepCopyInto(&newTempRes)
+	labels := newTempRes.GetLabels()
+	labels[flag.LabelEnvironmentType] = kp.environment
+	newTempRes.SetLabels(labels)
+	name := newTempRes.GetName()
+	newTempRes.SetName(name + kp.nameSuffix)
+
+	taskDeploy, found, err := unstructured.NestedFieldNoCopy(newTempRes.Object, "spec", "tasks")
+	if !found || err != nil {
+		return unstructured.Unstructured{}, err
+	}
+	taskDeployMap := taskDeploy.([]interface{})[2].(map[string]interface{})
+	taskDeployMap["name"] = "kn-service-create"
+	taskDeployMap["taskRef"] = map[string]interface{}{"name": "kn", "kind": "ClusterTask"}
+	taskDeployMap["runAfter"] = []interface{}{"build"}
+
+	taskDeployMap["params"] = []interface{}{
+		map[string]interface{}{"name": "ARGS", "value": []interface{}{"service", "create", "$(params.APP_NAME)", "--image=$(params.IMAGE_NAME)", "--force"}},
+	}
+	return newTempRes, nil
+}
+
+func CreatePipelines(template mf.Manifest, client client.Client) (mf.Manifest, error) {
+	var pipelines []unstructured.Unstructured
+
+	pipelineGenerators := []pipelineGenerator{
+		&dcPipeline{environment: "openshift", nameSuffix: ""},
+		&deploymentPipeline{environment: "kubernetes", nameSuffix: "-deployment"},
+		&knativePipeline{environment: "knative", nameSuffix: "-knative"},
+	}
```

Comment:
`{environment: "openshift", nameSuffix: ""}` 
it is the same struct.

what we need here is a single struct,
and 
which takes in 
- enviorment
- nameSuffix
- and an interface type object which defines the difference in the pipelines